assembly/src/tokens/lines.rs

```diff
@@ -0,0 +1,449 @@
+use super::{Token, Vec};
+use core::str::Lines;
+
+// LINES STREAM
+// ================================================================================================
+
+#[derive(Debug, Clone)]
+pub struct LinesStream<'a> {
+    lines: Lines<'a>,
+    count: u32,
+}
```

Comment:
I actually liked the previous implementation better (just with token handling removed). I would have modified it like this:
```Rust
#[derive(Debug, Clone)]
pub struct LineStream<'a> {
    lines: Lines<'a>,
    current_line: Option<&'a str>,
    current_line_num: u32,
    line_char_offset: u32,
}

impl<'a> From<&'a str> for LineStream<'a> {
    fn from(contents: &'a str) -> Self {
        Self {
            lines: contents.lines(),
            current_line: None,
            current_line_num: 0,
            line_char_offset: 0,
        }
    }
}

impl<'a> LineStream<'a> {
    /// Returns true if the current line is a token.
    fn is_token(&self) -> bool {
        self.current_line
            .filter(|line| !line.is_empty() && !line.starts_with(Token::LINE_COMMENT_PREFIX))
            .is_some()
    }

    /// Returns true if the current line is a token or a doc comment.
    fn is_token_or_doc_comment(&self) -> bool {
        self.current_line
            .filter(|line| {
                !line.is_empty() && !line.starts_with(Token::LINE_COMMENT_PREFIX)
                    || line.starts_with(Token::DOC_COMMENT_PREFIX)
            })
            .is_some()
    }

    /// If the current line is a doc comment, take lines until EOF or not doc comment.
    fn take_docs_block(&mut self) -> Vec<&'a str> {
        iter::from_fn(|| {
            self.current_line
                .and_then(|line| line.strip_prefix(Token::DOC_COMMENT_PREFIX))
                .map(|doc| doc.trim())
                .map(|doc| {
                    self.go_to_next_line();
                    doc
                })
        })
        .fold(Vec::with_capacity(10), |mut v, doc| {
            if !doc.trim().is_empty() {
                v.push(doc)
            }
            v
        })
    }

    /// Move the pointer to the next line, updating the control variables
    fn go_to_next_line(&mut self) {
        self.current_line = self.lines.next();
        if let Some(line) = self.current_line {
            let init_len = line.len();
            let trimmed = line.trim_start();

            self.current_line.replace(trimmed);
            self.line_char_offset = (init_len - trimmed.len()) as u32;
            self.current_line_num += 1;
        }
    }
}

impl<'a> Iterator for LineStream<'a> {
    type Item = LineInfo<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        self.go_to_next_line();

        // read next line and halt if empty
        while !self.is_token_or_doc_comment() {
            self.go_to_next_line();
            self.current_line?;
        }

        // fetch a docs block, returning if not followed by a token
        let docs = self.take_docs_block();
        if !docs.is_empty() && !self.is_token() {
            let line = if self.current_line.is_none() {
                self.current_line_num
            } else {
                self.current_line_num.saturating_sub(1)
            };
            return Some(LineInfo::new(line).with_docs(docs));
        }

        // seek next token
        while !self.is_token() {
            self.go_to_next_line();
            self.current_line?;
        }

        // fetch current line
        match self.current_line {
            Some(line) => Some(
                LineInfo::new(self.current_line_num)
                    .with_contents(line, self.line_char_offset)
                    .with_docs(docs),
            ),
            None => {
                debug_assert!(false, "this is unreachable; these is a bug in `Self::is_token`");
                None
            }
        }
    }
}
```

Reply:
Added as experiment here: https://github.com/0xPolygonMiden/miden-vm/pull/865/commits/2430969e5dd261a43aca84dc324e548ae077ec3f

We however have some open decisions:

- `add mul # foo` is returning `add mul` on the previous implementation, and `add mul # foo` on the proposed implementation
- given the previous point, we can no longer distinguish postfix dangling docs from a regular line, so we leave this responsibility for `TokenStream`

Reply:
> * `add mul # foo` is returning `add mul` on the previous implementation, and `add mul # foo` on the proposed implementation

We could strip this too but i'm not sure it is needed. The component which handles tokenization of individual lines could handle it in the future.

> * given the previous point, we can no longer distinguish postfix dangling docs from a regular line, so we leave this responsibility for `TokenStream`

This is relevant for something like this - right?:
```
begin
  push.1
  push.2
  add
end #! some comment
```
I think we should create an issue to address this in the future as we actually don't handle end of line doc comments correctly anyway now.